hpx/hpx_init_impl.hpp

```diff
@@ -296,137 +381,97 @@ namespace hpx
 
-        return init(main_f, desc_commandline, argc, argv, cfg,
-            empty, empty, mode);
-    }
-
-    // Main entry point for launching the HPX runtime system.
-    inline int
-    init(util::function_nonser<int(int, char**)> const& f,
-        int argc, char** argv, hpx::runtime_mode mode)
-    {
-        std::string app_name(HPX_APPLICATION_STRING);
-        return init(f, app_name, argc, argv, mode);
+        hpx::init_params iparams;
+        iparams.desc_cmdline = desc_cmdline;
+        iparams.mode = mode;
+        return init(f, argc, argv, iparams);
     }
 
     inline int
     init(util::function_nonser<int(int, char**)> const& f,
         int argc, char** argv, std::vector<std::string> const& cfg,
         hpx::runtime_mode mode)
     {
-        std::string app_name(HPX_APPLICATION_STRING);
-        using hpx::program_options::options_description;
-
-        options_description desc_commandline(
-            "Usage: " + app_name +  " [options]");
-
-        util::function_nonser<int(hpx::program_options::variables_map& vm)>
-            main_f = util::bind_back(detail::init_helper, f);
-
         HPX_ASSERT(argc != 0 && argv != nullptr);
 
-        return init(main_f, desc_commandline, argc, argv, cfg,
-            startup_function_type(), shutdown_function_type(), mode);
+        hpx::init_params iparams;
+        iparams.cfg = cfg;
+        iparams.mode = mode;
+        return init(f, argc, argv, iparams);
     }
 
     inline int
         init(util::function_nonser<int(int, char**)> const& f,
         std::vector<std::string> const& cfg,
         hpx::runtime_mode mode)
     {
-        char *dummy_argv[2] = { const_cast<char*>(HPX_APPLICATION_STRING), nullptr };
-
-        return init(f, 1, dummy_argv, cfg, mode);
+        hpx::init_params iparams;
+        iparams.cfg = cfg;
+        iparams.mode = mode;
+        return init(f, 1, iparams.dummy_argv, iparams);
     }
 
     inline int
     init(std::nullptr_t, std::string const& app_name, int argc, char** argv,
         hpx::runtime_mode mode)
     {
         using hpx::program_options::options_description;
-        options_description desc_commandline(
-            std::string("Usage: ") + app_name +  " [options]");
+        options_description desc_cmdline("Usage: " + app_name +  " [options]");
 
-        util::function_nonser<int(hpx::program_options::variables_map& vm)>
-            main_f;
-        std::vector<std::string> cfg;
-        util::function_nonser<void()> const empty;
+        util::function_nonser<int(hpx::program_options::variables_map&)> main_f;
 
         HPX_ASSERT(argc != 0 && argv != nullptr);
 
-        return init(main_f, desc_commandline, argc, argv, cfg,
-            empty, empty, mode);
+        hpx::init_params iparams;
+        iparams.desc_cmdline = desc_cmdline;
+        iparams.mode = mode;
+        return init(main_f, argc, argv, iparams);
     }
 
     inline int
     init(std::nullptr_t const& f,
```

Comment:
```suggestion
    init(std::nullptr_t f,
```
(my fault!). Could you please remove the `const&`?